components/wifi/esp8266-driver/ESP8266/ESP8266.cpp

```diff
@@ -612,9 +614,18 @@ bool ESP8266::dns_lookup(const char *name, char *ip)
     return done;
 }
 
-nsapi_error_t ESP8266::send(int id, const void *data, uint32_t amount)
+nsapi_size_or_error_t ESP8266::send(int id, const void *data, uint32_t amount)
 {
+    if (_prev_send_ok_pending && _sock_i[id].proto == NSAPI_TCP) {
+        tr_debug("send(): Previous packet was not ACK-ed with SEND OK.");
+        return NSAPI_ERROR_WOULD_BLOCK;
+    }
```

Comment:
Is it necessary to trigger `ATCmdParser` to roll, like calling `process_oob`? Otherwise, OOB message `SEND OK` is not handled.

Reply:
@ccli8 , please correct me if I am wrong, but it seems that sigio will automatically trigger oob processing.

`ESP8266`'s sigio is [connected to serial's sigio](https://github.com/ARMmbed/mbed-os/blob/5f495db06bd6c6e0ac0adf001be57f1d724f9aa1/components/wifi/esp8266-driver/ESP8266/ESP8266.cpp#L1001).
`ESP8266Interface`'s sigio is [connected to that of](https://github.com/ARMmbed/mbed-os/blob/5f495db06bd6c6e0ac0adf001be57f1d724f9aa1/components/wifi/esp8266-driver/ESP8266Interface.cpp#L132) `ESP8266`.
So if anything comes over serial it will trigger `ESP8266Interface::event`, [which in turn will schedule](https://github.com/ARMmbed/mbed-os/blob/5f495db06bd6c6e0ac0adf001be57f1d724f9aa1/components/wifi/esp8266-driver/ESP8266Interface.cpp#L1028) `ESP8266Interface::proc_oob_evnt()` (unless there is one scheduled already).
That [will trigger oob processing](https://github.com/ARMmbed/mbed-os/blob/5f495db06bd6c6e0ac0adf001be57f1d724f9aa1/components/wifi/esp8266-driver/ESP8266Interface.cpp#L1124).

Does this make sense?

Reply:
@michalpasztamobica Thanks for your explanation. That makes sense.

Still have other concern with it. When some exception like disconnect or error happens, `_prev_send_ok_pending` doesn't get cleared and program will block here indefinitely. With `_prev_send_ok_pending` and `_send_fail_received` mutual exclustion, plus other exception cases with send ack, recommend combining `_prev_send_ok_pending` and `_send_fail_received` into one which has states like: no send, send pending, send ok acked, and send fail acked. One point is that program cannot block here even in send ack exception cases.

```C++
    if (_prev_send_ok_pending && _sock_i[id].proto == NSAPI_TCP) {
        tr_debug("send(): Previous packet was not ACK-ed with SEND OK.");
        return NSAPI_ERROR_WOULD_BLOCK;
    }
```

Potential memory leak with new oobs:
```C++
_parser.oob("SEND FAIL", callback(this, &ESP8266::_oob_send_fail_received));
_parser.oob("SEND OK", callback(this, &ESP8266::_oob_send_ok_received));
```
`_parser.oob` is not easy to pair `_parser.remove_oob` in exception cases. Recommend just placing them, like other oobs, in `ESP8266::ESP8266` constructor and modifying related logic.